api/src/convert/amount.rs

```diff
@@ -16,18 +35,62 @@ impl From<&MaskedAmount> for external::MaskedAmount {
     }
 }
 
-impl TryFrom<&external::MaskedAmount> for MaskedAmount {
+impl From<&VersionedMaskedAmount> for external::TxOut_oneof_masked_amount {
+    fn from(source: &VersionedMaskedAmount) -> Self {
+        match source {
+            VersionedMaskedAmount::V1(masked_amount) => {
+                external::TxOut_oneof_masked_amount::v1(masked_amount.into())
+            }
+            VersionedMaskedAmount::V2(masked_amount) => {
+                external::TxOut_oneof_masked_amount::v2(masked_amount.into())
+            }
+        }
+    }
+}
+
+impl TryFrom<&external::MaskedAmount> for MaskedAmountV1 {
     type Error = ConversionError;
 
     fn try_from(source: &external::MaskedAmount) -> Result<Self, Self::Error> {
         let commitment = CompressedCommitment::try_from(source.get_commitment())?;
         let masked_value = source.get_masked_value();
         let masked_token_id = source.get_masked_token_id();
-        let amount = MaskedAmount {
+        let amount = MaskedAmountV1 {
             commitment,
             masked_value,
             masked_token_id: masked_token_id.to_vec(),
         };
         Ok(amount)
     }
 }
+
+impl TryFrom<&external::MaskedAmount> for MaskedAmountV2 {
+    type Error = ConversionError;
+
+    fn try_from(source: &external::MaskedAmount) -> Result<Self, Self::Error> {
+        let commitment = CompressedCommitment::try_from(source.get_commitment())?;
+        let masked_value = source.get_masked_value();
+        let masked_token_id = source.get_masked_token_id().to_vec();
+        let amount = MaskedAmountV2 {
+            commitment,
+            masked_value,
+            masked_token_id,
+        };
+        Ok(amount)
+    }
+}
+
+impl TryFrom<&external::TxOut_oneof_masked_amount> for VersionedMaskedAmount {
```

Comment:
If we can invent around the challenges with this, this seems like the superior approach.